Adds priv_getcode

josh-richardson opened the pull request:
priv_getCode was previously implemented on another branch, but that became stale & difficult to rebase, so it was quicker to just branch off master and pull in my relevant commits to this branch by cherry-picking.


Signed-off-by: Joshua Richardson <joshua@richardson.tech>

<!-- Thanks for sending a pull request! Please check out our contribution guidelines: -->
<!-- https://github.com/hyperledger/besu/blob/master/CONTRIBUTING.md -->

## PR description
Adds priv_getCode, the implementation of eth_getCode but for private transactions.

## Fixed Issue(s)
<!-- Please link to fixed issue(s) here using format: fixes #<issue number> -->
<!-- Example: "fixes #2" -->

lucassaldanha commented on ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetCode.java:
```diff
@@ -0,0 +1,93 @@
+
+  public PrivGetCode(
+      final BlockchainQueries blockchain,
+      final PrivacyParameters privacyParameters,
```
It looks like we are passing down the `PrivacyParameter` object only to get an instance of the `privateStateArchive`. I think we should change this to receive the privateStateArchive directly.

lucassaldanha commented on ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetCode.java:
```diff
@@ -0,0 +1,93 @@
+
+  @Override
+  public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
+    LOG.trace("Executing {}", RpcMethod.PRIV_GET_CODE.getMethodName());
```
We already have DEBUG log level for JSON-RPC calls. I don't think we need this one.

lucassaldanha commented on ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetCode.java:
```diff
@@ -0,0 +1,93 @@
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+
+public class PrivGetCode implements JsonRpcMethod {
```
We have an abstraction for methods that have the block paramater to handle named or numbers, etc.
Take a look at `PrivCall` and how it extends `AbstractBlockParameterMethod`. I believe we should make `PrivGetCode` extend `AbstractBlockParameterMethod`.

lucassaldanha commented on ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetCode.java:
```diff
@@ -0,0 +1,93 @@
+  public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
+    LOG.trace("Executing {}", RpcMethod.PRIV_GET_CODE.getMethodName());
+
+    final Address address =
```
We have standardized that for any priv_ methods that have an equivalent eth_ method, we will make the first parameter (index 0) the privacyGroupId. And all other parameters will be the same.

So, the expected params for `priv_getCode` are:
```
params: [
   'privacy_group_id',
   '0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b',
   '0x2'  // 2
]
```

Take a look at `PrivCall` implementation

lucassaldanha commented on ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetCode.java:
```diff
@@ -0,0 +1,93 @@
+    final Hash latestStateRoot =
+        privateStateRootResolver.resolveLastStateRoot(
+            privacyGroupId,
+            blockParameter.isNumeric() && blockParameter.getNumber().isPresent()
```
As I mentioned before, extending `AbstractBlockParameterMethod` should help you handling the block paramter.

lucassaldanha commented:
Would you please add an entry on `CHANGELOG.md` under the `1.4.1` section mentioning that we are adding `priv_getCode`?

josh-richardson commented:
I've now addressed all comments from @lucassaldanha and @pinges, however by changing the parameter order, Web3j has had to undergo a refactor which has not yet been released, and therefore the `DeployPrivateSmartContractAcceptanceTest` will currently fail due to Web3j dependency sending the params in the incorrect order. I hope for this to be fixed on Monday & to be able to make another commit updating the Web3j dependency.

josh-richardson commented:
Bumped the Web3j version, so this should now be ready to go.

lucassaldanha commented on ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetCodeTest.java:
```diff
@@ -0,0 +1,145 @@
+            new JsonRpcRequest(
+                "2.0",
+                "priv_getCode",
+                new Object[] {
```
We need to fix the order of the parameters here.
